p4app-exercises/p4runtime.p4app/README.md

```diff
@@ -1,174 +1,183 @@
-actions, as defined in the P4Info file produced by the compiler (look for the
-file `build/basic.p4info` after executing `make run`). Any changes in the P4
-program that add or rename tables, keys, or actions will need to be reflected in
-these `sX-runtime.json` files.
-
-## Step 2: Implement L3 forwarding
-
-The `basic.p4` file contains a skeleton P4 program with key pieces of
-logic replaced by `TODO` comments. Your implementation should follow
-the structure given in this file---replace each `TODO` with logic
-implementing the missing piece.
-
-A complete `basic.p4` will contain the following components:
-
-1. Header type definitions for Ethernet (`ethernet_t`) and IPv4 (`ipv4_t`).
-2. **TODO:** Parsers for Ethernet and IPv4 that populate `ethernet_t` and `ipv4_t` fields.
-3. An action to drop a packet, using `mark_to_drop()`.
-4. **TODO:** An action (called `ipv4_forward`) that:
-	1. Sets the egress port for the next hop. 
-	2. Updates the ethernet destination address with the address of the next hop. 
-	3. Updates the ethernet source address with the address of the switch. 
-	4. Decrements the TTL.
-5. **TODO:** A control that:
-    1. Defines a table that will read an IPv4 destination address, and
-       invoke either `drop` or `ipv4_forward`.
-    2. An `apply` block that applies the table.   
-6. **TODO:** A deparser that selects the order
-    in which fields inserted into the outgoing packet.
-7. A `package` instantiation supplied with the parser, control, and deparser.
-    > In general, a package also requires instances of checksum verification
-    > and recomputation controls. These are not necessary for this tutorial
-    > and are replaced with instantiations of empty controls.
+within each table are inserted by the control plane. In this case,
+`mycontroller.py` implements our control plane, instead of installing static
+table entries like we have in the previous exercises.
+
+**Important:** A P4 program also defines the interface between the
+switch pipeline and control plane. This interface is defined in the
+`advanced_tunnel.p4info` file. The table entries that you build in `mycontroller.py`
+refer to specific tables, keys, and actions by name, and we use a P4Info helper
+to convert the names into the IDs that are required for P4Runtime. Any changes
+in the P4 program that add or rename tables, keys, or actions will need to be
+reflected in your table entries.
+
+## Step 2: Implement Tunnel Forwarding
+
+The `mycontroller.py` file is a basic controller plane that does the following:
+1. Establishes a gRPC connection to the switches for the P4Runtime service.
+2. Pushes the P4 program to each switch.
+3. Writes tunnel ingress and tunnel egress rules for two tunnels between h1 and h2.
+4. Reads tunnel ingress and egress counters every 2 seconds.
+
+It also contains comments marked with `TODO` which indicate the functionality
+that you need to implement.
```

Comment:
Could we make it clear that students can incrementally edit `mycontroller.py` and run `make mycontroller to test their changes, without needing to run `make run` again?

Reply:
I can add a statement to that effect somewhere in there, in my next commit.